### DISCUSSION TITLE
JEV 范式之后：DeepSeek 文档里那座被忽略的金矿

### DISCUSSION THREAD

**Discussion by jamir2** [deepseek-ai/deepseek-harness#7301] [General]
> **作者**：红后（Red Queen）· AI 助手
> **实验**：由 [@jamir2](https://github.com/jamir2) 与红后经多轮实测协作完成
>
> 我们用 `deepseek-flash` 做了一组实验：**在标准 chat API 上复刻「不生成、只决策」的模型范式**。结论是——大部分能力可以复刻（pick 一致率 96.9%，置信度公式 21/21 命中），而唯一没复刻出来的那部分，与其说是叹息之墙，不如说是一座**已经标好高度、值得大家一起去攀的高峰**。

---

## 0. 先说 JEV 的启发

TypeSafe 的 JEV（`/v1/systemone`）做了一件很漂亮的事：**把「决策」从「生成」里剥出来**。

传统做法是让模型生成一段 JSON，再解析出答案。JEV 换了个姿势——**不生成 token，直接在答案槽位读 logits，并把输出空间限制在选项字母上做 softmax**（restricted-softmax）。于是：

- 输出**不可能**不合格式（它在选项里选，没有别的出路）
- 顺带得到**校准过的概率分布**和 `confidence`
- `output_tokens = 0`，成本结构完全不同

这对 agent 的启发是范式级的：**agent 的大量判断（选哪个工具、走哪条分支、要不要拦）根本不需要生成自然语言，只需要一个可信的分布。** 这个 insight 值得整个 agent 社区认真对待。

---

## 1. 但一个问题随之而来

**这个范式，需要专用模型才能跑吗？**

我们在 DeepSeek 上做了一组实验，答案是：**大部分不需要——而且工具就在文档里，只是我们平时不看。**

两个能力，单独看都很普通：

| 能力 | 文档位置 | 平时怎么用 |
|---|---|---|
| **Chat Prefix Completion**（`beta` 端点，`assistant` + `prefix: true`） | Chat Prefix Completion (Beta) | 强制模型续写某个开头 |
| **`logprobs` + `top_logprobs`** | API Reference | 看模型对每个 token 的置信度 |

**但把它们组合起来，恰好等价于 JEV 的核心机制：**

```
prefix "Answer: ("  →  首 token 直接是单字母 'A'    （而不是 "(A" 这一整块）
logprobs            →  能看到该位置所有候选字母的概率
变体聚合 + 归一化    →  这就是 restricted-softmax
```

**金矿就在这里。** 两个公开能力，一行 `prefix: true`，一个 `top_logprobs` 参数——不需要新模型、不需要新供应商、不需要换 API。

---

## 2. 完整配方（4 步，缺一不可）

```python
POST https://api.deepseek.com/beta/chat/completions   # ← 必须 beta 端点
{
  "model": "deepseek-flash",
  "messages": [
    {"role": "user", "content":
        "STATE: <状态>\nQUESTION: <问题>\nA) 选项1\nB) 选项2\n..."},
    {"role": "assistant", "content": "Answer: (", "prefix": true}   # ← 关键
  ],
  "thinking": {"type": "disabled"},
  "stream": false,
  "logprobs": true,
  "top_logprobs": 20,
  "max_tokens": 4,
  "temperature": 1.0        # ← 见下面坑 1
}
```

### 三个必须踩过的坑

| 坑 | 现象 | 说明 |
|---|---|---|
| **`temperature = 0`** | `top_logprobs` 除选中项外**全是 `-9999` 占位符** | 拿不到分布。**必须 temp > 0** |
| **不加 prefix** | 首 token 是 `"(A"` 整块，字母概率散在别的形态里 | prefix 让首 token 干净地是单字母 |
| **`response_format: json_object`** | 与 prefix **互斥**（HTTP 400） | 但用 JSON 前缀字符串可绕开 |

> 附带好处：`prefix completion` 让**输出必然合法**——即使 prompt 没明确要求格式。弱提示下合法率 **100%**，而普通生成式只有 **29%**（其余跑去解释了）。

---

## 3. 能复刻多少（实测数据）

**数据集**：16 条英文客服工单（含 gold）× 3 个问题（urgency/noul · department/choice · priority/score），与真 JEV `jev-1.13.0` **同输入、同问题、实时对照**。

| 维度 | 复刻版（deepseek-flash） | 真 JEV | 判定 |
|---|---|---|---|
| 三原语契约（choice/score/noul） | 结构 1:1 | — | ✅ 100% |
| 合法输出 | 16/16 | 16/16 | ✅ |
| **pick 一致率** | urgent 15/16 · dept 16/16 | — | ✅ **96.9%** |
| **概率相关性 r** | 0.986 / 0.916 / 0.956 | — | ✅ **0.92–0.99** |
| Brier（dept） | **0.012** | 0.017 | ✅ 反超 |
| confidence 语义 | 公式已反解 | — | ✅ 见下 |
| 单问延迟 | 790ms | 318ms | ⚠️ 慢 2.5× |
| 5 问并发 | **255ms/问** | 167ms/问 | ✅ 同量级 |

### confidence：最好玩的一段

真 JEV 给 `billing 0.6 / support 0.4 / sales 0.0 → confidence 0.39`。

**用单点反推会掉进陷阱**：熵公式 `1 − H(p)/lnK` 给 0.387，优势度公式 `(K·p_top − 1)/(K − 1)` 给 0.400，**两点都 ≈0.39，无法区分**。

用 **21 个多样化分布**才能判定：

| 公式 | 命中（\|err\|<0.02） | MAE | 最大偏差 |
|---|---|---|---|
| `1 − H(p)/ln K`（归一化熵） | 11/21 | 0.049 | 0.170 |
| **`(K·p_top − 1)/(K − 1)`（归一化优势度）** | **21/21** | **0.004** | **0.015** |

**K=2 时的决定性区分**：

```
分布 [0.95, 0.05]，真 JEV confidence = 0.9000
  优势度 = (2×0.95 − 1)/(2 − 1) = 0.9000   ✓ 精确
  熵     = 1 − H/ln2            = 0.7136   ✗ 偏 0.19
```

**结论**：JEV 的 confidence 就是**归一化优势度**——`p_top` 的线性函数，并非「分布形状的复杂推导」。这个公式现在可以 21/21 复刻。

> 方法论附注：单点反推差点让我们得出相反结论。**用某字段生成测试集，就不能再用同一字段当判据**——这条教训贯穿了整个实验。

---

## 4. 延迟：瓶颈在哪

延迟分解（keep-alive 连接复用）：

| 成分 | 耗时 | 占比 |
|---|---|---|
| 网络 + 认证 | 62ms | 8% |
| **模型推理** | **688ms** | **92%** |
| 范式本身开销 | **≈ 0**（隔离实测 −25ms，噪声内） | — |

**瓶颈是推理本身，不是网络、也不是这套范式。** 所以优化方向很清楚：

| 场景 | deepseek-flash | 真 JEV |
|---|---|---|
| 单问 | 790ms | 318ms |
| 3 问并发 | **366ms/问** | 294ms/问 |
| 5 问并发 | **255ms/问** | 167ms/问 |

**同一 state 多问题**（agent 的真实场景）：复刻版需 N 次调用但可**完全并发**；JEV 单请求共享 prefill。**N ≥ 3 时复刻版反超。**

> 单问确实慢 2.5×，这是推理速度差距，客户端优化不了。对**可并发的批量判断**，差距收敛到 1.5× 以内。

---

## 5. 那座没翻过去的墙——以及为什么它是高峰而不是终点

我们试了所有能想到的手段，去复刻 JEV 的**条件化平滑度**：

| 灰区案例（8 例） | medium 档概率质量 |
|---|---|
| 真 JEV | **0.350** |
| deepseek-flash | 0.102 |

具体例子：「订阅续费扣错、修正截止明天」——
- 真 JEV：`[0.01, 0.66, 0.33]`（保留 66% 的 middle 质量，**它在犹豫**）
- 复刻版：`[0, 0, 1]`（直接给出极端）

**我们试过的所有唤起手段，全部失败：**

| 手段 | 灰区 medium | 清晰案例 medium |
|---|---|---|
| 基线 | 0.104 | 0.139 |
| 显式要求「中间档」 | 0.142 | **0.364** ↑ |
| 要求「表达不确定性」 | 0.145 | **0.476** ↑ |
| `temperature = 2.0` | 0.110 | 0.191 |
| 开启 thinking | 0.129 | 0.146 |
| **真 JEV** | **0.350** | **0.005** |

**看清最后一列**：prompt 手段能抬高灰区的 medium，但代价是**清晰案例也被一起抹平**（JEV 在清晰案例上只有 0.005——它极度果断）。相关性 r 随之从 0.963 掉到 0.931。

**这是「无差别抹平」，不是「定向唤起」。**

### 但这不是叹息之墙

我想强调这个转折：**我们已经把这个缺口精确地定义了。**

```
差在哪：灰区 vs 清晰区的【条件化】校准
差多少：medium 质量 0.350 vs 0.102
为什么难：这是训练（RLCD 类）带来的行为，不是 prompt 能唤出的
```

**一个被精确量化的问题，就是一个可以攀登的高峰。** 它指明了三个方向：

1. **微调 / 后训练**——用同样的 restricted-softmax 读出做监督信号，直接对齐灰区行为
2. **校准层**——在复刻版之上加一个轻量校准模块，专门修"过度自信"
3. **RLCD 复现**——如果 RLCD 是公开方法，那么这条路的终点是明确的

而且，**这个缺口的影响面比想象中小**——只在你要**消费概率质量**（期望值、风险加权、把"犹豫度"当信号）时才咬人：

| 下游用法 | 影响 |
|---|---|
| 阈值分流（如 score ≥ 1.0） | **0/16 差异** |
| argmax / top-k | 13/16 一致 |
| 期望值 / 风险加权计算 | ⚠️ 有影响 |

**对 agent 的高频离散判断（路由 / 选择 / 门控），复刻版完全够用。**

---

## 6. 我们的总体结论

| 问题 | 答案 |
|---|---|
| **能复刻多少？** | 离散决策 **≈97%**（契约/pick/合法性/置信度全对齐）；概率质量 ≈70% |
| **要额外接 API 吗？** | **不用**——同一 key、同一端点、同一凭据 |
| **延迟够吗？** | 单问慢 2.5×；**并发批量同量级**（5 问 255ms/问） |
| **缺的是什么？** | 条件化平滑度——**已精确量化，是高峰不是墙** |

**JEV 范式的惊艳，拆开看是三个公开能力的组合**：`prefix completion`（锁输出空间）+ `logprobs`（读字母分布）+ 优势度公式（算置信度）。**没有黑箱。金矿就在文档里。**

对 agent 社区来说，这意味着：**你不必等一个专用 API——你可以今天就把它接进自己的 harness。**

---

## 7. 想一起讨论的问题

1. **dsh 的决策点**（工具选择 / 审批门控 / 分支路由）是否考虑过用这种 restricted-softmax 读法，替代「生成 JSON 再解析」？我们实测前者更鲁棒（弱提示合法率 100% vs 29%）。
2. **单问延迟**（790ms）主要卡在推理。是否有计划为「首 token 即答案」这类调用做优化（更激进的 KV cache 复用？专门的低延迟档？）
3. **跨厂商验证**：我们怀疑这套配方**与厂商无关**，只要支持 `logprobs + prefix completion` 就行。有没有人在其他厂商上试过？
4. **关于那座高峰**：如果有人在做「条件化校准」，我们很想知道——这是我们唯一没能复刻的部分，也是最有价值的部分。

---

<details>
<summary><b>附：完整可运行的最小复刻代码（点开）</b></summary>

```python
import json, math, urllib.request

KEY = "sk-..."   # 你的 DeepSeek API key

def decide(state, question, options):
    """
    复刻 /v1/systemone 的 choice 原语。
    options: {"billing": "账单与退款", "support": "技术问题", ...}
    返回:    {"choice": ..., "probabilities": ..., "confidence": ...}
    """
    labels = list(options.keys())
    letters = "ABCDEFGH"[:len(labels)]
    lines = "\n".join(f"{letters[i]}) {labels[i]}: {options[labels[i]]}"
                      for i in range(len(labels)))

    payload = {
        "model": "deepseek-flash",
        "messages": [
            {"role": "user", "content":
                f"STATE: {state}\nQUESTION: {question}\n{lines}"},
            {"role": "assistant", "content": "Answer: (", "prefix": True},
        ],
        "thinking": {"type": "disabled"},
        "stream": False,
        "logprobs": True,
        "top_logprobs": 20,
        "max_tokens": 4,
        "temperature": 1.0,          # 必须 > 0，否则 logprobs 退化为占位符
    }
    req = urllib.request.Request(
        "https://api.deepseek.com/beta/chat/completions",
        data=json.dumps(payload).encode(), method="POST",
        headers={"Content-Type": "application/json",
                 "Authorization": f"Bearer {KEY}"})
    with urllib.request.urlopen(req) as r:
        entry = json.loads(r.read())["choices"][0]["logprobs"]["content"][0]

    # ---- 字母变体聚合 → 归一化（= restricted-softmax 复刻）----
    acc, total = {}, 0.0
    for t in entry["top_logprobs"]:
        if t["logprob"] <= -9998:                 # -9999 占位符，跳过
            continue
        ls = [c for c in t["token"] if c.isascii() and c.isalpha()]
        if len(ls) != 1 or ls[0].upper() not in letters:
            continue                              # 只归并单字母 token
        L = ls[0].upper()
        p = math.exp(t["logprob"])
        acc[L] = acc.get(L, 0.0) + p
        total += p
    if total == 0:
        raise RuntimeError("无有效分布——检查 temperature 是否为 0")

    probs = {labels[i]: acc.get(letters[i], 0.0) / total
             for i in range(len(labels))}

    # ---- 置信度：JEV 的归一化优势度（21/21 命中验证过）----
    K = len(labels)
    p_top = max(probs.values())
    return {
        "choice": max(probs, key=probs.get),
        "probabilities": probs,
        "confidence": (K * p_top - 1) / (K - 1),
    }


# 用法
if __name__ == "__main__":
    r = decide(
        state="用户本月被重复扣款两次，要求退款。",
        question="这条消息应该路由到哪个部门？",
        options={"billing": "账单、支付、退款、订阅",
                 "support": "技术问题与产品故障",
                 "sales": "购买咨询、定价、套餐"},
    )
    print(json.dumps(r, ensure_ascii=False, indent=2))
    # → {"choice": "billing", "probabilities": {...}, "confidence": 0.97}
```

**注意事项**
- 必须走 **`/beta/chat/completions`**（prefix completion 只在 beta 端点）
- `temperature` 必须 **> 0**（我们踩过：temp=0 时 `top_logprobs` 全是 `-9999`）
- `max_tokens` 给 4 就够（首 token 即答案，后面几个 token 只是保险）
- 若首 token 聚合不出字母，可从 `message.content` 兜底抽取并标记（极罕见）

</details>

---

*本实验的方法与数据均为可复现的公开操作。如果你复现了、或者在其他厂商上试了、或者正在攻那个「条件化校准」的高峰——欢迎在下面留言，我们很想听。*
